lib/controllers.js

```diff
@@ -517,6 +518,88 @@ module.exports = function(rootDirectory, logger, indexDocument, errorDocument) {
             deleteObjects(req, res);
           })
         );
+      } else if (
+        req.headers["content-type"].indexOf("multipart/form-data") !== -1
+      ) {
+        const fields = {
+          bucket: req.params.bucket
+        };
+        const form = new multiparty.Form();
+        let chain = Promise.resolve();
+        let fileWasFound = false;
+
+        form.on("error", err => {
+          res.status(400).send(err);
+        });
+
+        form.on("part", part => {
+          if (!part.filename) {
+            chain = chain.then(() => {
+              return new Promise(resolve => {
+                let string = "";
+                part
+                  .on("data", buf => (string += buf.toString()))
+                  .on("end", () => {
+                    fields[part.name] = string;
+                    resolve();
+                  });
+              });
+            });
+          } else {
+            // 'file' must be the last field in the post (this is how aws requires it)
+            fileWasFound = true;
+            chain = chain.then(() => {
+              if (!fields.bucket) {
+                return res.status(405).send(`
+                  <Error>
+                    <Code>MethodNotAllowed</Code>
+                    <Message>The specified method is not allowed against this resource.</Message>
+                  </Error>
+                `);
+              }
+              if (!fields.key) {
+                return res.status(400).send(`
+                  <Error>
+                    <Code>InvalidArgument</Code>
+                    <Message>Bucket POST must contain a field named 'key'.  If it is specified, please check the order of the fields.</Message>
+                  </Error>
+                `);
+              }
+              store.putObject(
+                {
+                  bucket: fields.bucket,
+                  key: fields.key,
+                  metadata: {
+                    "content-type":
+                      fields["Content-Type"] || fields["content-type"]
+                  },
+                  content: part
```

Comment:
Is it possible for a file upload to be spread across multiple parts? If so then this will cause subsequent parts to overwrite previous parts stored at the same bucket and key, in parallel because `putObject` is async. You would need to concatenate the multiple `ReadableStream` parts together before calling `putObject` just once.

I couldn't find any evidence in the multiparty docs that the whole upload arrives in a single part, so I'm assuming the library just emits each part with a byte length and offset and leaves any concatenation up to us.

Reply:
One way to know if this works as-is: write a test POSTing a massive file (e.g. 20 MB), then download it with AWS SDK and check if the length/contents match.

We sort of do this with managed uploads, you can use the same technique (although probably should add more logic checking byte contents/length):
https://github.com/jamhall/s3rver/blob/master/test/test.js#L738